### DISCUSSION TITLE
Cannot link custom search extensibility library in custom webpart solution

### DISCUSSION THREAD

**Discussion by arunakshi** [microsoft-search/pnp-modern-search#2362] [Q&A] [answered]
I have a custom search extensibility library solution where we have lot of search results cards displayed via PnP modern search webpart. I would like to reuse the same components (react components which will be rendered from WebComponents) in my custom webpart since we have a requirement to display the same cards in custom webpart along with other content. 

I tried using npm link custom-search-extensibility-library in my custom webpart solution. It failed to reference because of deep dependency reference as my custom search extensibility library is linking to another custom library solution. 

I am able to link custom search extensibility library in my custom webpart solution with the help of npm install ./custom-search-extensibility-library. However, I am getting following when I try to render the React Card Components in my custom webpart. I end up having the following error: **Uncaught (in promise) Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.**

I checked all my imports from custom webpart solution and exports from custom search extensibility library solution, everything looks fine. 

Does anyone tried this way to reuse the custom search extensibility library components.

**Comment by FranckyC** [comment, marked as answer]
Hi @arunakshi, first of all, make sure, the SPFx and TypeScript versions are the same between all projects. For your specific use case, it looks lkie you'll need some code refactoring, like isolate the React controls in a dedicated library and then link it to both extensibility library and your custom Web Part.

**Reply by arunakshi** [reply]
Hi FranckyC, Thanks for your prompt reply. I re-created my custom webpart with same version of SPFX as in my custom search extensibility library. However, I am facing same issue. 

Our client do not want to go with a dedicated library for reach controls since we have already 3 other library solutions. Is it supported to link extensibility library in custom webpart. I haven't found the official documentation for it.

**Reply by FranckyC** [reply]
The extensibility library is a regular SPFx library project so nothing particular here. You should be able to link it. Did you export your control correctly in the library in the `index.ts` file?

**Reply by arunakshi** [reply]
Thanks alot FranckyC.  Its working now! I exported the components from search extensibility library correctly, strangely I have to reference that react component in my webpart solution individually like this: `import {CardComponent} from 'custom-search-extensibility/lib/Components/CardComponent/CardComponent';`

I use to reference my linked library components like this (see below) and its working always in my other solutions. 

```
import * as customLibrary from 'custom-favs-library';
<customLibrary.CardComponent {...props}/>
```
Somehow, the second way of import not working for me when I linked search extensibility library in my custom webpart solution. we can live with that individual component reference.

**Reply by FranckyC** [reply]
Glad it works now ! It is always better to used named exports and import only the components you really need in your code ;)
